src/LocalDatastore.js

```diff
@@ -38,44 +56,69 @@ const LocalDatastore = {
     return controller.getAllContents();
   },
 
+  // Use for testing
+  _getRawStorage(): Promise {
+    const controller = CoreManager.getLocalDatastoreController();
+    return controller.getRawStorage();
+  },
+
   _clear(): Promise {
     const controller = CoreManager.getLocalDatastoreController();
     return controller.clear();
   },
 
   // Pin the object and children recursively
   // Saves the object and children key to Pin Name
-  async _handlePinWithName(name: string, object: ParseObject): Promise {
+  async _handlePinAllWithName(name: string, objects: Array<ParseObject>): Promise {
+    if (!objects || objects.length === 0) {
+      return;
+    }
     const pinName = this.getPinName(name);
-    const objects = this._getChildren(object);
-    objects[this.getKeyForObject(object)] = object._toFullJSON();
-    for (const objectKey in objects) {
-      await this.pinWithName(objectKey, objects[objectKey]);
+    const promises = [];
+    const objectKeys = [];
+    for (const parent of objects) {
+      const children = this._getChildren(parent);
+      const parentKey = this.getKeyForObject(parent);
+      children[parentKey] = parent._toFullJSON();
+      for (const objectKey in children) {
+        objectKeys.push(objectKey);
+        promises.push(this.pinWithName(objectKey, children[objectKey]));
+      }
     }
+    await Promise.all(promises);
     const pinned = await this.fromPinWithName(pinName) || [];
```

Comment:
would be good to use 

```
const pinPromises = this.fromPinWithName(pinName) || promise.resolve([]);
const objectKeyPromises = Promise.all(promises);
const [objectKeys, pinned] = await Promises.all([objectKeyPromises, pinPromises]);
```
so that both operations happen concurrently.  As you have it, they go serially even though they are not dependent on each other.

disclaimer, my code above is likely just mostly right as it would take some more restructuring.

unfortunately, i think this should be fixed.